### PR TITLE
Add `System.perf_counter/{0,1}`

### PR DESCRIPTION
**Pull request by mtrudel**
As [discussed on elixir-lang-core](https://groups.google.com/u/1/g/elixir-lang-core/c/uzPDT6Y2ziA), this PR adds support for `System.perf_counter/{0,1}` intended for use in the context of metric & telemetry gathering.

### CODE REVIEW

**Comment by josevalim** [comment]
Thanks but, given the very precise use case, IMO it is completely fine to reach out to OTP for this. :)

**Comment by mtrudel** [comment]
Sounds good!

**Comment by halostatue** [comment]
This might not be a bad thing to add in Telemetry documentation. One of the weaknesses I have found with Telemetry is the lack of "best practices" documentation. Whether the os of `:os.perf_counter/{0,1}` is a best practice or not, better documentation around those sorts of trade-offs with telemetry would be good.

**Comment by mtrudel** [comment]
Agreed - the intent of this PR was around discoverability & documentation more than enablement. Telemetry et al may be a better home for this discoverability.

**Comment by mtrudel** [comment]
It's mostly that on the one side, OTP provides a tool for the *express purpose* of performance measurement, and on the Elixir / lib side there's a ton of cargo culting around the use of `system_time` for this purpose. I've been writing Telemetry code for a long time and didn't know about `perf_counter`; it just feels like it should be more widely known about / used / promoted as a best practice.

**Comment by josevalim** [comment]
Is there a reason then why Erlang's own `timer:tc/1` does not use `os:perf_counter()`? I will be glad to incorporate this function if Erlang itself uses this for its own measurements.

Although `telemetry` itself is hard to change, because everything assumes native unit.

**Comment by mtrudel** [comment]
I've noticed and wondered that too @josevalim. My best guess is that `perf_counter` came out in OTP 19, whereas `timer:tc` dates from R14. 

There are also valid places where `monotonic_time` is more suitable for purpose, given its guarantees. The way I've come to think about is if measurement is a fundamental part of your domain (as it would be for something like `timer::tc`, then you should be using `monotonic_time` to make *absolutely sure* that your measurements are correct. On the other hand, if measurement is more adjunct to your domain, as telemetry usually is to your core logic, then `perf_counter` makes more sense since it trades iron-clad accuracy for a much lower footprint. It's a bit of a subtle difference. 

Looking at OTP's implementation, `perf_counter` ends up calling right into whatever TSC register the underlying CPU has (if it has one); for example it will call to RDTSC directly on x86: https://github.com/erlang/otp/blob/82d134f1b4132209337a5c58d3b8115816b725d0/erts/emulator/sys/unix/sys_time.c#L1015. So `perf_counter` is absolutely as fast as you're ever gonna get. 

Moreover, the numbers don't really lie - quoted from my original email on the list:

```
Mix.install([{:benchee, "~> 1.0", only: :dev}])

Benchee.run(%{
  system_time: &System.system_time/0,
  monotonic_time: &System.monotonic_time/0,
  perf_counter: &:os.perf_counter/0
})

Name                     ips        average  deviation         median         99th %
perf_counter         54.03 M       18.51 ns   ±439.35%       18.33 ns       20.01 ns
monotonic_time       28.10 M       35.58 ns  ±2758.64%       33.30 ns       41.70 ns
system_time          23.44 M       42.67 ns  ±8964.02%       37.50 ns       45.90 ns

Comparison:
perf_counter         54.03 M
monotonic_time       28.10 M - 1.92x slower +17.08 ns
system_time          23.44 M - 2.31x slower +24.16 ns
```

tl;dr is I'm quite sure that `perf_counter` is what people doing telemetry should be calling 99% of the time (people doing timing as a core part of their domain are in a more complicated situation though). Getting the community to steer in this direction is pretty exclusive of the work here though, other than (perhaps) the visibility of having ready access & discoverability within stdlib.

**Comment by halostatue** [comment]
> Although `telemetry` itself is hard to change, because everything assumes native unit.

Note that I’m not suggesting that `telemetry` itself change. In most cases, people implementing telemetry need to provide their own payloads. Many libraries just use `:erlang.monotonic_time/{0,1}`, since that’s what (IIRC) `:telemetry.span` does and what most other extant clients do. What I’m talking about is just a matter of adding some documentation on "best practices for implementing telemetry".

I’m not volunteering to write such, because I’m not sure of the best practices as I’ve only implemented telemetry in a single library and not in an application as of yet. We might be able to get @am-kantox (telemetria) or @vereis (sibyl) to collaborate on such, but I think that @mtrudel has raised a good point that for *some cases*, `:os.perf_counter` is likely to be more performant than `:erlang.montonic_time`…but being able to distinguish those cases might be desirable.

All that said, this discussion should probably be taking place somewhere in beam-telemetry/telemetry.
